src/engine/math_util.c

```diff
@@ -1490,3 +1490,85 @@ void find_surface_on_ray(Vec3f orig, Vec3f dir, struct Surface **hit_surface, Ve
         }
     }
 }
+
+#include <world_scale.h>
+
+// Constructs a float in registers, which can be faster than gcc's default of loading a float from rodata.
+// Especially fast for halfword floats, which get loaded with a `lui` + `mtc1`.
+static ALWAYS_INLINE float construct_float(const float f)
+{
+    u32 r;
+    float f_out;
+    u32 i = *(u32*)(&f);
+
+    if (!__builtin_constant_p(i))
+    {
+        return *(float*)(&i);
+    }
+
+    u32 upper = (i >> 16);
+    u32 lower = (i >>  0) & 0xFFFF;
+
+    if ((i & 0xFFFF) == 0) {
+        __asm__ ("lui %0, %1"
+                                : "=r"(r)
+                                : "K"(upper));
+    } else if ((i & 0xFFFF0000) == 0) {
+        __asm__ ("addiu %0, $0, %1"
+                                : "+r"(r)
+                                : "K"(lower));
+    } else {
+        __asm__ ("lui %0, %1"
+                                : "=r"(r)
+                                : "K"(upper));
+        __asm__ ("addiu %0, %0, %1"
+                                : "+r"(r)
+                                : "K"(lower));
+    }
+
+    __asm__ ("mtc1 %1, %0"
+                         : "=f"(f_out)
+                         : "r"(r));
+    return f_out;
+}
+
+// Converts a floating point matrix to a fixed point matrix
+// Makes some assumptions about certain fields in the matrix, which will always be true for valid matrices.
+__attribute__((optimize("Os")))
+void mtxf_to_mtx_fast(s16* dst, float* src)
+{
+    float scale = construct_float(65536.0f / WORLD_SCALE);
+    // Iterate over pairs of values in the input matrix
+    for (int i = 0; i < 8; i++)
+    {
+        // Read the first input in the current pair
+        float a = src[2 * i + 0];
+
+        // Convert the first input to fixed
+        s32 a_int = (s32)(a * scale);
+        dst[2 * i +  0] = (s16)(a_int >> 16);
+        dst[2 * i + 16] = (s16)(a_int >>  0);
+
+        // If this is the left half of the matrix, convert the second input to fixed
+        if ((i & 1) == 0)
+        {
+            // Read the second input in the current pair
+            float b = src[2 * i + 1];
+            s32 b_int = (s32)(b * scale);
+            dst[2 * i +  1] = (s16)(b_int >> 16);
+            dst[2 * i + 17] = (s16)(b_int >>  0);
+        }
+        // Otherwise, skip the second input because column 4 will always be zero
+        // Row 4 column 4 is handled after the loop.
+        else
+        {
+            dst[2 * i +  1] = 0;
+            dst[2 * i + 17] = 0;
+        }
+
+    }
+    // Write 1.0 to the bottom right entry in the output matrix
+    // The low half was already set to zero in the loop, so we only need
+    //  to set the top half.
+    dst[15] = 1;
+}
```

Comment:
newline at end of file?
```suggestion
}

```